### ISSUE TITLE
Rapid memory leak on Linux

### ISSUE THREAD

**Issue by m00dawg** [smeighan/xLights#879]
I was importing a sequence from last year which has some elements that do not exist in this year's show. Once finished, when I went to re-render the scene, I ran into a very rapid memory leak (it chewed through 32GB of RAM in less than 10 seconds). I tried re-mapping the elements that I could as well as ignoring the elements and just straight up deleting the elements. All seem to cause the leak.

Since not all of my old sequences cause this, I've included two - Carol of the Bells (which has the issue) and Sandstorm (which doesn't).

[bug.tar.gz](https://github.com/smeighan/xLights/files/1117228/bug.tar.gz)

**Comment by cjd** [comment]
I just tried to load the two sequences - but none of the models used in the sequences exist in the show layout.
What models/groups are they meant to be remapped to?
Also - can you please grab the /tmp/xLights_l4cpp.log file for me?
Finally - which version/build of xLights are you using?

**Comment by m00dawg** [comment]
Hmm that's weird. Both may complain about a few missing models (by design so you can compare the two) although it should only be a few models. For Carol of the Bells, the missing ones are:

Full Roof -> Pixel Roof
Tree Branches -> Tree Globes
Roof - Lower Right -> Removed
Roof - Lower Left -> Removed
Steeples - Right -> Removed
Steeples - Left -> Removed

I tried both remapping the full roof and tree branches as well as deleting them (I can easily resequence them) but both caused the bug.

I tried to grab the log but didn't find it in /tmp whether or not I use the -d flag. Is there some other means I should be trying to generate that?

Version of xLights is the github trunk (commit f2f16438d89d820fa74dea450f9f0c8302dd741a)

**Comment by cjd** [comment]
I remapped those models and saw the same behaviour. I am tracking down what is causing it currently

**Comment by cjd** [comment]
Narrowed down to 'Pixel Tree' model effects - getting there

**Comment by cjd** [comment]
This doesn't seem to be specific to linux as I can reproduce in windows as well
I've narrowed it down to the single-strand effects on rows within the pixel tree
Attached is a cut-down sequence with just two single-strand effects - and it takes 330 seconds to render using 4G of memory (every effect added increases memory dramatically) for such a simple sequence.
[problem.zip](https://github.com/smeighan/xLights/files/1118413/problem.zip)
If I get rid of the SubModelEffectLayer section of the pixeltree from the original sequence (removing all the effects on the rows in the pixel tree) it renders fine - so that is the root of the issue

**Comment by cjd** [comment]
The problem is with the submodels for the pixeltree.
If you delete the existing submodels then re-generate them it gets past the error and works once more

**Comment by cjd** [comment]
How were the submodels created originally?
They are overlapping, go outside the buffer and some are zero sized (which seems a bit odd)

**Comment by m00dawg** [comment]
Ah I assume you mean the Row 1 - Row 25 models? I created them in the Layout section under the Pixel Tree options then went to Sub Models. There I basically manually created each row and filtered it.

I did this so I could do individual effects on rows rather than columns of the pixel tree easily. Makes sense that this sequence is the one that breaks since the other sequence doesn't have any effects on the rows.

I now see there is an option to generate rows (YEAH!!!! THAT'S FANTASTIC!) - at the time that optino wasn't available so I did it manually. Oh man now that I know that's there, that's going to be wonderful!
